pep795.md

```diff
@@ -0,0 +1,1938 @@
+   immutable objects. This takes pressure off normal cycle 
+   detection in both subinterpreters and free-threading, and
+   is key to sharing objects between subinterpreters without
+   needing substantial changes to cycle detection.
+
+In this PEP, we strive to create a semantics for immutable
+objects that is consistent across both subinterpreters and
+free-threaded Python. This is a difficult problem as
+subinterpreters rely on isolation plus a per-interpreter GIL for
+correctness, whereas free-threading has no GIL.
+
+
+# Semantics of Immutability
+
+This section describes the immutability semantics. All user-facing
+functions in this PEP are in the module `immutable`. The most
+important function is `freeze()` that makes an object immutable.
+Note that freezing is *in-place*; `freeze(obj)` freezes `obj` and
+for convenience also returns a reference to (now immutable) `obj`.
+
+```python
+class Person:
+    def __init__(self, name):
+        self.name = name
+
+monty = [Person("Eric"), Person("Graham"), Person("Terry")]
+freeze(monty)
+
+monty.append(Person("John")) # throws exception because monty is immutable
+monty.pop()                  # --''--
+
+monty[0].name = "John"       # throws exception because all person objects 
+                             # in monty are immutable too
+```
+
+Making an object immutable means that subsequent attempts
+to write to its fields will raise an exception. (The attempt
+to append or pop above will both result in `TypeError:
+object of type list is immutable`.) Furthermore, by default,
+immutability is **deep**. This means that an immutable
+object is only permitted to point to other immutable objects
+(modulo some escape hatches that will be described further
+down). This has consequences in Python as functions, modules
+and types are objects too. We will use the following running
+example to explain the semantics of freezing. We start by
+explaining the simplest and most permissive model of freezing,
+and later discuss how to limit how freezing may propagate
+through a program.
+
+```python
+def gcd(a, b):
+    while b:
+        a, b = b, a % b
+    return a
+
+def print_fraction(f):
+    counter += 1
+    print("Fraction: {f} (fractions printed in total: {counter})")
+
+class Fraction:
+    def __init__(self, numerator, denominator):
+        d = gcd(numerator, denominator)
+        self.n = numerator // d
+        self.d = denominator // d
+
+    def __add__(self, other):
+        num = self.n * other.d + other.n * self.d
+        den = self.d * other.d
+        return Fraction(num, den)
+
+    def __repr__(self):
+        return f"{self.n}/{self.d}"
+
+f1 = Fraction(1, 3)
+f2 = Fraction(2, 7)
+freeze(f1)
+f1 = f1 + f2
+```
+
+Let us first focus on "normal Python objects" that the
+user has explicitly created, meaning we ignore "declaration
+objects" such as modules, functions, and types.
+
+Right before freezing, we have the following object graph.
+
+```mermaid
```

Comment:
We should check if we can pin the theme to light theme, this looks weird in dark theme:

<img width="657" height="400" alt="Image" src="https://github.com/user-attachments/assets/ab229278-1b69-4b44-a284-08842f6e14cd" />

Reply:
Well spotted. I have addressed this so the rendering works in both light and dark theme. However, since the rST format does not support inline mermaid, we will have to commit to a style -- I guess?